File: csharp/Characters/TestCharacters/CharactersTest.cs
using Characters;
using NUnit;
using NUnit.Framework;

namespace TestCharacters;

public class CharactersTest
{
    public List<Character> allTestData()
    {
        var joyce = new Character("Joyce", "Byers");
        var jim = new Character("Jim", "Hopper");
        var mike = new Character("Mike", "Wheeler");
        var eleven = new Character("Eleven", null);
        var dustin = new Character("Dustin", "Henderson");
        var lucas = new Character("Lucas", "Sinclair");
        var nancy = new Character("Nancy", "Wheeler");
        var jonathan = new Character("Jonathan", "Byers");
        var will = new Character("Will", "Byers");
        var karen = new Character("Karen", "Wheeler");
        var steve = new Character("Steve", "Harrington");
        var mindflayer = new Character("Mindflayer", null, isMonster:true);
        var demagorgon = new Character("Demagorgon", null, isMonster:true);
        var demadog = new Character("Demadog", null, isMonster:true);
        
        joyce.AddChild(jonathan);
        joyce.AddChild(will);
        jim.AddChild(eleven);
        karen.AddChild(nancy);
        karen.AddChild(mike);
        
        eleven.AddNemesis(demagorgon);
        will.AddNemesis(mindflayer);
        dustin.AddNemesis(demadog);

        return new List<Character>()
        {
            joyce,
            jim,
            mike,
            will,
            eleven,
            dustin,
            lucas,
            nancy,
            jonathan,
            mindflayer,
            demagorgon,
            demadog,
            karen,
            steve,
        };
    }

    [TestCase]
    public void FindCharacterByFirstName()
    {
        var finder = new CharacterFinder(allTestData());
        var character = finder.FindByFirstName("Jim");
        Assert.AreEqual("Jim", character?.FirstName);
    }
}

File: scala/src/test/scala/characters/CharactersTest.scala
package characters

import org.scalatest.OptionValues.*
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers.*

class CharactersTest extends AnyFunSuite:

  private def allTestData: List[Character] =
    val joyce = new Character("Joyce", Some("Byers"))
    val jim = new Character("Jim", Some("Hopper"))
    val mike = new Character("Mike", Some("Wheeler"))
    val eleven = new Character("Eleven")
    val dustin = new Character("Dustin", Some("Henderson"))
    val lucas = new Character("Lucas", Some("Sinclair"))
    val nancy = new Character("Nancy", Some("Wheeler"))
    val jonathan = new Character("Jonathan", Some("Byers"))
    val will = new Character("Will", Some("Byers"))
    val karen = new Character("Karen", Some("Wheeler"))
    val steve = new Character("Steve", Some("Harrington"))
    val mindflayer = new Character("Mindflayer", isMonster = true)
    val demagorgon = new Character("Demagorgon", isMonster = true)
    val demadog = new Character("Demadog", isMonster = true)

    joyce.addChild(jonathan)
    joyce.addChild(will)
    jim.addChild(eleven)
    karen.addChild(nancy)
    karen.addChild(mike)

    eleven.nemesis = demagorgon
    will.nemesis = mindflayer
    dustin.nemesis = demadog

    List[Character](
      joyce,
      jim,
      mike,
      will,
      eleven,
      dustin,
      lucas,
      nancy,
      jonathan,
      mindflayer,
      demagorgon,
      demadog,
      karen,
      steve
    )

  test("findCharactersByFirstName") {
    val finder = new CharacterFinder(allTestData)

    val character = finder.findByFirstName("Jim")

    character.value.firstName should be("Jim")
  }
